pkg/capabilities/consensus/ocr3/reporting_plugin.go

```diff
@@ -0,0 +1,232 @@
+	"github.com/smartcontractkit/chainlink-common/pkg/values"
+)
+
+var _ ocr3types.ReportingPlugin[any] = (*reportingPlugin)(nil)
+
+type responder interface {
+	response(ctx context.Context, resp response) error
+}
+
+type reportingPlugin struct {
+	batchSize int
+	s         *store
+	r         responder
+
+	// TODO: make this dynamic based on the request
+	aggregator pbtypes.Aggregator
+	encoder    pbtypes.Encoder
+}
+
+func newReportingPlugin(s *store, r responder, batchSize int) *reportingPlugin {
+	return &reportingPlugin{
+		s:         s,
+		r:         r,
+		batchSize: batchSize,
+	}
+}
+
+func (r *reportingPlugin) Query(ctx context.Context, outctx ocr3types.OutcomeContext) (types.Query, error) {
+	batch, err := r.s.firstN(ctx, r.batchSize)
+	if err != nil {
+		return nil, err
+	}
+
+	ids := []*pbtypes.RequestId{}
+	for _, r := range batch {
+		ids = append(ids, &pbtypes.RequestId{
+			RequestId:   r.RequestID,
+			NamespaceId: r.NamespaceID,
+		})
+	}
+
+	return proto.Marshal(&pbtypes.Query{
+		RequestIds: ids,
+	})
+}
+
+func (r *reportingPlugin) Observation(ctx context.Context, outctx ocr3types.OutcomeContext, query types.Query) (types.Observation, error) {
+	queryReq := &pbtypes.Query{}
+	err := proto.Unmarshal(query, queryReq)
+	if err != nil {
+		return nil, err
+	}
+
+	rids := []string{}
+	for _, q := range queryReq.RequestIds {
+		rids = append(rids, q.RequestId)
+	}
+
+	reqs, err := r.s.getN(ctx, rids)
+	if err != nil {
+		return nil, err
+	}
+
+	obs := &pbtypes.Observation{}
+	for _, r := range reqs {
+		obsPb, err := r.Observations.Proto()
+		if err != nil {
+			// TODO: we can't process this observation: log and continue
+		}
+
+		r := &pbtypes.Request{
+			Observation: obsPb,
+			RequestId: &pbtypes.RequestId{
+				RequestId:   r.RequestID,
+				NamespaceId: r.NamespaceID,
+			},
+		}
+
+		obs.Requests = append(obs.Requests, r)
+	}
+
+	return proto.Marshal(obs)
+}
+
+func (r *reportingPlugin) ValidateObservation(outctx ocr3types.OutcomeContext, query types.Query, ao types.AttributedObservation) error {
+	// TODO : just try and unmarshal the message
+	// basic validation
+	return nil
+}
+
+func (r *reportingPlugin) ObservationQuorum(outctx ocr3types.OutcomeContext, query types.Query) (ocr3types.Quorum, error) {
+	return ocr3types.QuorumTwoFPlusOne, nil
+}
+
+func (r *reportingPlugin) Outcome(outctx ocr3types.OutcomeContext, query types.Query, aos []types.AttributedObservation) (ocr3types.Outcome, error) {
+	m := map[string]map[ocrcommon.OracleID][]values.Value{}
+	for _, o := range aos {
+		obs := &pbtypes.Observation{}
+		err := proto.Unmarshal(o.Observation, obs)
+		if err != nil {
+			// TODO log
+			continue
+		}
+
+		for _, r := range obs.Requests {
+			rid := r.RequestId.RequestId
+			if _, ok := m[rid]; !ok {
+				m[rid] = make(map[ocrcommon.OracleID][]values.Value)
+			}
+
+			val, err := values.FromProto(r.Observation)
+			if err != nil {
+				continue
+			}
+
+			m[rid][o.Observer] = append(m[rid][o.Observer], val)
```

Comment:
Could we rename "rid" to workflowID? Those request IDs are pretty confusing when reading.

Reply:
Yes, will do! I think I was over-optimising a bit here